src/main/java/sparksoniq/jsoniq/runtime/iterator/postfix/PredicateIterator.java

```diff
@@ -176,7 +181,22 @@ private void setNextResult() {
         RuntimeIterator iterator = this._children.get(0);
         RuntimeIterator filter = this._children.get(1);
         JavaRDD<Item> childRDD = iterator.getRDD(dynamicContext);
-        if(filter.getVariableDependencies().containsKey("$position") || filter.getVariableDependencies().containsKey("last"))
+        if(
+        		!filter.getVariableDependencies().containsKey("$position")
+        		&& !filter.getVariableDependencies().containsKey("last")
+        		&& (
+        			filter instanceof BooleanRuntimeIterator
+        			|| filter instanceof AndOperationIterator
+        			|| filter instanceof OrOperationIterator
+        			|| filter instanceof NotOperationIterator
+        			|| filter instanceof ComparisonOperationIterator
+        		)
+        ) {
+        	Function<Item, Boolean> transformation = new PredicateClosure(filter, dynamicContext);
```

Comment:
This is the optimized (unzipped part) if positions are not needed. Runs as before.